api/clients/v2/payload_retriever.go

```diff
@@ -2,283 +2,12 @@ package clients
-	if err != nil {
-		return fmt.Errorf(
-			"generate and compare commitment for blob %v received from relay %v: %w",
-			blobKey.Hex(),
-			relayKey,
-			err)
-	}
-
-	if !valid {
-		return fmt.Errorf("commitment for blob %v is invalid for bytes received from relay %v", blobKey.Hex(), relayKey)
-	}
-
-	// Checking that the length returned by the relay is <= the length claimed in the BlobCommitments is sufficient
-	// here: it isn't necessary to verify the length proof itself, since this will have been done by DA nodes prior to
-	// signing for availability.
-	//
-	// Note that the length in the commitment is the length of the blob in symbols
-	if uint(len(blob)) > blobLength*encoding.BYTES_PER_SYMBOL {
-		return fmt.Errorf(
-			"length for blob %v (%d bytes) received from relay %v is greater than claimed blob length (%d bytes)",
-			blobKey.Hex(),
-			len(blob),
-			relayKey,
-			blobLength*encoding.BYTES_PER_SYMBOL)
-	}
-
-	return nil
-}
-
-// getBlobWithTimeout attempts to get a blob from a given relay, and times out based on config.RelayTimeout
-func (pr *PayloadRetriever) getBlobWithTimeout(
-	ctx context.Context,
-	relayKey core.RelayKey,
-	blobKey core.BlobKey) ([]byte, error) {
-
-	timeoutCtx, cancel := context.WithTimeout(ctx, pr.config.RelayTimeout)
-	defer cancel()
-
-	return pr.relayClient.GetBlob(timeoutCtx, relayKey, blobKey)
-}
-
-// verifyCertWithTimeout verifies an EigenDACert by making a call to VerifyCertV2.
-//
-// This method times out after the duration configured in payloadRetrieverConfig.ContractCallTimeout
-func (pr *PayloadRetriever) verifyCertWithTimeout(
-	ctx context.Context,
-	eigenDACert *verification.EigenDACert,
-) error {
-	timeoutCtx, cancel := context.WithTimeout(ctx, pr.config.ContractCallTimeout)
-	defer cancel()
-
-	return pr.certVerifier.VerifyCertV2(timeoutCtx, eigenDACert)
-}
-
-// Close is responsible for calling close on all internal clients. This method will do its best to close all internal
-// clients, even if some closes fail.
-//
-// Any and all errors returned from closing internal clients will be joined and returned.
-//
-// This method should only be called once.
-func (pr *PayloadRetriever) Close() error {
-	err := pr.relayClient.Close()
-	if err != nil {
-		return fmt.Errorf("close relay client: %w", err)
-	}
-
-	return nil
-}
-
-// blobCommitmentsBindingToInternal converts a blob commitment from an eigenDA cert into the internal
-// encoding.BlobCommitments type
-func blobCommitmentsBindingToInternal(
-	blobCommitmentBinding *verifiercontract.BlobCommitment,
-) (*encoding.BlobCommitments, error) {
-
-	blobCommitment, err := encoding.BlobCommitmentsFromProtobuf(
-		verification.BlobCommitmentBindingToProto(blobCommitmentBinding))
-
-	if err != nil {
-		return nil, fmt.Errorf("blob commitments from protobuf: %w", err)
-	}
-
-	return blobCommitment, nil
+// PayloadRetriever represents something that knows how to retrieve a payload from some backend using a verification.EigenDACert
```

Comment:
Explain why we need more than one backend: relay, nodes, and potentially s3 are the 3 main candidates

Reply:
Btw does our RetrievalClient implement this interface? Should we not modify it as part of this PR?

Reply:
[95b2a3ee](https://github.com/Layr-Labs/eigenda/pull/1187/commits/95b2a3ee7c432b4de22eb618aa54a0459ed2ae3a)

Reply:
> Btw does our RetrievalClient implement this interface? Should we not modify it as part of this PR?

I don't think so. Just as we have the `RelayPayloadRetriever`, which wraps around the `RelayClient`, I think we need to have a `DaNodePayloadRetriever`, which wraps around the `RetrievalClient`.

Reply:
Ideally, the an end user would have a single payload retriever that knows how to talk to relays and validators. The payload retriever would make a best effort attempt to fetch the data from the relays, only falling back to the validators if unsuccessful at getting the data from the relays.

Reply:
> Ideally, the an end user would have a single payload retriever that knows how to talk to relays and validators.

Each individual rollup may have different configurations for how they want to retrieve payloads. Should you retry from relays before trying a fallback? Should you fallback to some S3 bucket instead of falling back to DA nodes? If our only wrapper for payload retrieval is opinionated here, then that limits options.

Practically speaking, the main "user" of this utility is our own proxy. Most "end users" will mainly interact with the proxy. But having this interface also lets more advanced end users provide the proxy with their own retrieval backends, if required.

Reply:
Good discussion. Resolving.